server/src/main/java/keywhiz/service/resources/BatchSecretDeliveryResource.java

```diff
@@ -0,0 +1,150 @@
+    this.clientDAO = clientDAOFactory.readonly();
+  }
+
+  @VisibleForTesting
+  BatchSecretDeliveryResource(SecretController secretController, AclDAO aclDAO,
+                              ClientDAO clientDAO) {
+    this.secretController = secretController;
+    this.aclDAO = aclDAO;
+    this.clientDAO = clientDAO;
+  }
+
+  /**
+   * Retrieve Secret by name
+   *
+   * @param secrets the name of the Secrets to retrieve in batch
+   * @param client  the client performing the retrieval
+   * @return the secret with the specified name, if present and accessible to the client
+   * <p>
+   * responseMessage 200 Found and retrieved Secret with given name
+   * responseMessage 403 Secret is not assigned to Client
+   * responseMessage 404 Secret with given name not found
+   * responseMessage 500 Secret response could not be generated for given Secret
+   */
+  @Timed
+  @ExceptionMetered
+  @POST
+  @Consumes(APPLICATION_JSON)
+  public List<SecretDeliveryResponse> getBatchSecret(@Auth Client client, @Valid BatchSecretRequest request) {
+
+
+    List<SanitizedSecret> clientAccessibleSecrets = aclDAO.getBatchSanitizedSecretsFor(client, request.secrets());
+    List<Secret> existingSecrets = secretController.getSecretsByName(request.secrets());
+
+    boolean clientExists = clientDAO.getClientByName(client.getName()).isPresent();
+
+    List<String> forbiddenSecrets = new ArrayList<>();
+
+    // The request fails whenever a single secret is requested that is not accessible
+    // The client is responsible for only requesting secrets they have permission for
+    for (String secretname : request.secrets()) {
+      boolean secretExists = existingSecrets.stream().anyMatch(s -> s.getName().equals(secretname));
+      boolean secretAccessible = clientAccessibleSecrets.stream().anyMatch(s -> s.name().equals(secretname));
+
+      if (!secretAccessible) {
+        if (clientExists && secretExists) {
+          forbiddenSecrets.add(secretname);
+        } else {
+          if (clientExists) {
+            logger.info("Client {} requested unknown secret {}", client.getName(), secretname);
+          }
+          throw new NotFoundException();
+        }
+      } else {
+        if (!secretExists) {
+          // If the secret is accessible but does not exist,
+          // this indicates a Keywhiz issue. This is not a client issue.
+          logger.warn("Client {} requested secret which is not retrievable {}", client.getName(), secretname);
+          throw new NotFoundException("Secret not found.");
+        }
+      }
+    }
+
+    // If *any* of the secrets is forbidden
+    if ( ! forbiddenSecrets.isEmpty() ) {
+      throw new ForbiddenException(format("Access denied: %s to secret(s) '%s'", client.getName(), forbiddenSecrets));
+    }
+
+    logger.info("Client {} granted access to {}.", client.getName(), clientAccessibleSecrets.stream().map(s -> s.name()).collect(toList()));
+    try {
+      return existingSecrets.stream()
```

Comment:
nit: It could be worth either explicitly returning `clientAccessibleSecrets` or adding a comment here to clarify that we can't reach this code unless `clientAccessibleSecrets` and `existingSecrets` are identical--I initially thought this could return existing secrets that the client didn't have access to.

Reply:
I added a comment. Summary offline discussion: `clientAccessibleSecrets`are SanitizedSecrets, that change wouldn't be possible.